substrate/primitives/runtime/src/proving_trie.rs

```diff
@@ -0,0 +1,178 @@
+use sp_trie::{
+	trie_types::{TrieDBBuilder, TrieDBMutBuilderV0},
+	LayoutV0, MemoryDB, Recorder, Trie, TrieMut, EMPTY_PREFIX,
+};
+
+/// A trait for creating a merkle trie for checking and generating merkle proofs.
+pub trait ProvingTrie<Hashing, Hash, Key, Value>
+where
+	Self: Sized,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>;
+	/// Access the underlying trie root.
+	fn root(&self) -> &Hash;
+	/// Check a proof contained within the current memory-db. Returns `None` if the
+	/// nodes within the current `MemoryDB` are insufficient to query the item.
+	fn query(&self, key: Key) -> Option<Value>;
+	/// Create the full verification data needed to prove a key and its value in the trie. Returns
+	/// `None` if the nodes nodes within the current `MemoryDB` are insufficient to create a proof.
+	fn create_proof(&self, key: Key) -> Option<Vec<Vec<u8>>>;
+	/// Create a new instance of `ProvingTrie` from raw nodes. Nodes can be generated using the
+	/// `create_proof` function.
+	fn from_nodes(root: Hash, nodes: &[Vec<u8>]) -> Self;
+}
+
+/// A basic trie implementation for checking and generating proofs for a key / value pair.
+pub struct BasicProvingTrie<Hashing, Hash, Key, Value>
+where
+	Hashing: sp_core::Hasher<Out = Hash>,
+{
+	db: MemoryDB<Hashing>,
+	root: Hash,
+	_phantom: core::marker::PhantomData<(Key, Value)>,
+}
+
+impl<Hashing, Hash, Key, Value> ProvingTrie<Hashing, Hash, Key, Value>
+	for BasicProvingTrie<Hashing, Hash, Key, Value>
+where
+	Hashing: sp_core::Hasher<Out = Hash>,
+	Hash: Default + Send + Sync,
+	Key: Encode,
+	Value: Encode + Decode,
+{
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>,
+	{
+		let mut db = MemoryDB::default();
+		let mut root = Default::default();
+
+		{
+			let mut trie = TrieDBMutBuilderV0::new(&mut db, &mut root).build();
+			for (key, value) in items.into_iter() {
+				key.using_encoded(|k| value.using_encoded(|v| trie.insert(k, v)))
+					.map_err(|_| "failed to insert into trie")?;
+			}
+		}
+
+		Ok(Self { db, root, _phantom: Default::default() })
+	}
+
+	fn root(&self) -> &Hash {
+		&self.root
+	}
+
+	fn query(&self, key: Key) -> Option<Value> {
+		let trie = TrieDBBuilder::new(&self.db, &self.root).build();
+		key.using_encoded(|s| trie.get(s))
+			.ok()?
+			.and_then(|raw| Value::decode(&mut &*raw).ok())
+	}
+
+	fn create_proof(&self, key: Key) -> Option<Vec<Vec<u8>>> {
+		let mut recorder = Recorder::<LayoutV0<Hashing>>::new();
+
+		{
+			let trie =
+				TrieDBBuilder::new(&self.db, &self.root).with_recorder(&mut recorder).build();
+
+			key.using_encoded(|k| {
+				trie.get(k).ok()?.and_then(|raw| Value::decode(&mut &*raw).ok())
+			})?;
+		}
+
+		Some(recorder.drain().into_iter().map(|r| r.data).collect())
+	}
+
+	fn from_nodes(root: Hash, nodes: &[Vec<u8>]) -> Self {
```

Comment:
this function need a warning in documentation stating that this only work for small set of content (proving on big set like air drop, this would be plain terrible: we should need to map nodes of state individually or by group into the state and only access when needed the only node need by the proof).

Reply:
This is for checking proof isn't it? like by ensuring `from_nodes(proof).hash()` is equal to `root`

Reply:
Oh yes I am dumb  :+1: